lib/PipelineC/Tracing/Runner.cpp

```diff
@@ -32,13 +32,48 @@ static Logger<> TraceRunnerLogger("trace-runner");
 namespace utils {
 // Utility functions for decoding arguments
 
-// Put a nullptr at the end of the array and poison it, this will make it so
-// out-of-bounds errors are caught quickly
+// Utility class to store pointer arrays. When `.data()` is called, a nullptr is
+// added at the end of the array and it's poisoned by ASAN. This allows to spot
+// out of bounds accesses. After `.data()` has been called the internal vector
```

Comment:
ASan can already spot this kind of basic out of bounds accesses. [Godbolt](https://godbolt.org/z/98dqcnrYc)

Reply:
Hmmm, the general idea is that:
* Without ASAN, we put `nullptr` so that it's extra obvious if there is an OOB
* With ASAN, unless it's poisoned it would mask off-by-ones

Alternatively, I can:
* revert this and add a `#if __has_feature(address_sanitizer)` that omits adding the nullptr so that ASAN catches it
* drop the nullptr business entirely, but (IMHO) it might make those bugs harder to spot

Reply:
Ah, it's for when we *don't* have ASan, got it.

Can you persuade me that this particular piece of code is more worthy of the extra love compared to, say, any `std::vector<T *>` in the rest of the codebase?
Is this FUD or did you have a real scenario where this bit you?

Reply:
The origin of this implementation originates from the C implementation where padding the array with a `NULL` at the end was necessary to properly handle 0-length arrays. In this implementation it's recycled for 2 purposes:
* Have an obvious deference crash for out of bounds accesses
* Guarantee that even 0-sized arrays have a valid `.data()` pointer (in cppreference it says that it might be `nullptr` or it might be junk)

>  Is this FUD or did you have a real scenario where this bit you?

It didn't happen to me yet, but the spirit of `revng trace run` (in part) is to add extra checks so that easy-to-spot errors bubble up quickly (e.g. checking if the value of returns matches)

Reply:
OK